shared/stateutil/arrays.go

```diff
@@ -0,0 +1,158 @@
+package stateutil
+
+import (
+	"bytes"
+	"errors"
+	"sync"
+
+	"github.com/minio/highwayhash"
+	"github.com/protolambda/zssz/merkle"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+)
+
+var (
+	fastSumHashKey = bytesutil.ToBytes32([]byte("hash_fast_sum64_key"))
+	leavesCache    = make(map[string][][]byte)
+	layersCache    = make(map[string][][][]byte)
+	lock           sync.Mutex
+)
+
+func arraysRoot(roots [][]byte, fieldName string) ([32]byte, error) {
+	lock.Lock()
+	if _, ok := layersCache[fieldName]; !ok {
+		depth := merkle.GetDepth(uint64(len(roots)))
+		layersCache[fieldName] = make([][][]byte, depth+1)
+	}
+	lock.Unlock()
+
+	hashKeyElements := make([]byte, len(roots)*32)
+	leaves := make([][]byte, len(roots))
+	emptyKey := highwayhash.Sum(hashKeyElements, fastSumHashKey[:])
+	bytesProcessed := 0
+	changedIndices := make([]int, 0)
+	lock.Lock()
+	prevLeaves, ok := leavesCache[fieldName]
+	lock.Unlock()
+	for i := 0; i < len(roots); i++ {
+		copy(hashKeyElements[bytesProcessed:bytesProcessed+32], roots[i])
+		leaves[i] = roots[i]
+		// We check if any items changed since the roots were last recomputed.
+		if ok && prevLeaves != nil && !bytes.Equal(leaves[i], prevLeaves[i]) {
+			changedIndices = append(changedIndices, i)
+		}
+		bytesProcessed += 32
+	}
+
+	if len(changedIndices) > 0 {
+		var rt [32]byte
+		var err error
+		// If indices did change since last computation, we only recompute
+		// the modified branches in the cached Merkle tree for this state field.
+		chunks := leaves
+		for i := 0; i < len(changedIndices); i++ {
+			rt, err = recomputeRoot(changedIndices[i], chunks, fieldName)
+			if err != nil {
+				return [32]byte{}, err
+			}
+		}
+		return rt, nil
+	}
+
+	hashKey := highwayhash.Sum(hashKeyElements, fastSumHashKey[:])
```

Comment:
Is highway hash for sha256 faster than github.com/minio/sha256-simd? 

I think we should use only one and use `hashutil.Hash`